src/mdns/mdns_avahi.cpp

```diff
@@ -533,79 +663,174 @@ otbrError PublisherAvahi::PublishService(const char *   aHostName,
-    OTBR_UNUSED_VARIABLE(aType);
+    otbrError          error = OTBR_ERROR_NONE;
+    Services::iterator serviceIt;
+
+    VerifyOrExit(aName != nullptr, error = OTBR_ERROR_INVALID_ARGS);
+    VerifyOrExit(aType != nullptr, error = OTBR_ERROR_INVALID_ARGS);
+
+    serviceIt = FindService(aName, aType);
+    VerifyOrExit(serviceIt != mServices.end());
 
-    VerifyOrDie(false, "UnpublishService is not implemented with avahi");
+    otbrLog(OTBR_LOG_INFO, "[mdns] unpublish service %s.%s", aName, aType);
+    error = FreeGroup(serviceIt->mGroup);
+    mServices.erase(serviceIt);
 
-    return OTBR_ERROR_NONE;
+exit:
+    return error;
 }
 
 otbrError PublisherAvahi::PublishHost(const char *aName, const uint8_t *aAddress, uint8_t aAddressLength)
 {
-    OTBR_UNUSED_VARIABLE(aName);
-    OTBR_UNUSED_VARIABLE(aAddress);
-    OTBR_UNUSED_VARIABLE(aAddressLength);
+    otbrError       error      = OTBR_ERROR_NONE;
+    int             avahiError = 0;
+    Hosts::iterator hostIt     = mHosts.end();
+    char            fullHostName[kMaxSizeOfHost];
+    AvahiAddress    address;
+
+    VerifyOrExit(mState == State::kReady, errno = EAGAIN, error = OTBR_ERROR_ERRNO);
+    VerifyOrExit(mClient != nullptr, errno = EAGAIN, error = OTBR_ERROR_ERRNO);
+    VerifyOrExit(aName != nullptr, error = OTBR_ERROR_INVALID_ARGS);
+    VerifyOrExit(aAddress != nullptr, error = OTBR_ERROR_INVALID_ARGS);
+    VerifyOrExit(aAddressLength == sizeof(address.data.ipv6.address), error = OTBR_ERROR_INVALID_ARGS);
+    SuccessOrExit(error = MakeFullName(fullHostName, sizeof(fullHostName), aName));
+
+    hostIt = FindHost(aName);
+
+    if (hostIt == mHosts.end())
+    {
+        SuccessOrExit(error = CreateHost(*mClient, aName, hostIt));
+    }
+    else
+    {
+        // Check if the update is necessary
+        VerifyOrExit(memcmp(hostIt->mAddress.data.ipv6.address, aAddress, aAddressLength));
+        SuccessOrExit(error = ResetGroup(hostIt->mGroup));
+    }
 
-    VerifyOrDie(false, "PublishHost is not implemented with avahi");
+    address.proto = AVAHI_PROTO_INET6;
+    memcpy(&address.data.ipv6.address[0], aAddress, aAddressLength);
+
+    otbrLog(OTBR_LOG_INFO, "[mdns] create host %s", aName);
+    avahiError = avahi_entry_group_add_address(hostIt->mGroup, AVAHI_IF_UNSPEC, AVAHI_PROTO_INET6,
+                                               AVAHI_PUBLISH_NO_REVERSE, fullHostName, &address);
+    SuccessOrExit(avahiError);
+
+    otbrLog(OTBR_LOG_INFO, "[mdns] commit host %s", aName);
+    avahiError = avahi_entry_group_commit(hostIt->mGroup);
+    SuccessOrExit(avahiError);
+
+    hostIt->mAddress = address;
+
+exit:
+
+    if (avahiError)
+    {
+        error = OTBR_ERROR_MDNS;
+        otbrLog(OTBR_LOG_ERR, "Failed to publish host for avahi error: %s!", avahi_strerror(avahiError));
+    }
 
-    return OTBR_ERROR_NONE;
+    if (error == OTBR_ERROR_ERRNO)
+    {
+        otbrLog(OTBR_LOG_ERR, "Failed to publish host: %s!", strerror(errno));
+    }
```

Comment:
Avahi errors are handled a few lines above where the message explicitly states that it's an error from avahi. Without that it would require from a user to go through both otbr and avahi error messages to figure out where it comes from. Can we leave it as it is? :)

Reply:
If so, we won't show error messages for errors like `OTBR_ERROR_INVALID_ARGS` ? Thoughts?

Reply:
Ah, sorry, I must haven't woken up already and misunderstood your comment. I tried to stay consistent with the existing code which also didn't print errors for OTBR_ERROR_INVALID_ARGS etc. If you want me to change it, I will, but there are more places which require such a change.

Reply:
Ok, done.

Reply:
Thanks @Damian-Nordic ,

I am not sure if there is a convention of not printing `OTBR_ERROR_INVALID_ARGS`. But I think we should print all errors so that we don't miss any errors that could happen.

BTW, I found that `otbrErrorString` function already handled `OTBR_ERROR_ERRNO` differently, so we can make the code even simpler.